Userland/Applications/Assistant/main.cpp

```diff
@@ -206,29 +207,91 @@ class CalculatorProvider : public Provider {
     }
 };
 
+class AsyncProvider : public Provider {
+public:
+    void query(const String& query, Function<void(Vector<NonnullRefPtr<Result>>)> on_complete) override
+    {
+        (void)query;
+
+        if (m_timer)
+            return;
+
+        m_timer = Core::Timer::create_single_shot(1000, [&, on_complete = move(on_complete)] {
+            Vector<NonnullRefPtr<Result>> results;
+            results.append(adopt_ref(*new CalculatorResult("32")));
+            m_timer = nullptr;
+            on_complete(results);
+        });
+        m_timer->start();
+    }
+
+private:
+    RefPtr<Core::Timer> m_timer;
+};
+
 class Database {
 public:
+    explicit Database(AppState& state)
+        : m_state(state)
+    {
+    }
+
     Function<void(Vector<NonnullRefPtr<Result>>)> on_new_results;
 
     void search(const String& query)
     {
-        m_app_provider.query(query, [&](auto results) {
-            // FIXME: weigh and sort best results before calling 'on_new_results'
+        m_async_provider.query(query, [=, this](auto results) {
+            recv_results(query, results);
+        });
 
-            on_new_results(results);
+        m_app_provider.query(query, [=, this](auto results) {
+            recv_results(query, results);
         });
 
-        m_calculator_provider.query(query, [&](auto results) {
-            on_new_results(results);
+        m_calculator_provider.query(query, [=, this](auto results) {
+            recv_results(query, results);
         });
     }
 
 private:
+    void recv_results(const String& query, const Vector<NonnullRefPtr<Result>>& results)
+    {
+        {
+            Threading::Locker db_locker(m_lock);
+            auto it = m_result_cache.find(query);
+            if (it == m_result_cache.end()) {
+                m_result_cache.set(query, {});
+            }
+            it = m_result_cache.find(query);
+
+            for (auto& result : results) {
+                // FIXME: We probably want a better heuristic than title to determine if result exists
+                auto found = it->value.find_if([result](auto& other) {
+                    return result->title() == other->title();
+                });
+
+                if (found.is_end())
+                    it->value.append(result);
+            }
+        }
+
+        Threading::Locker state_locker(m_state.lock);
+        auto new_results = m_result_cache.find(m_state.last_query);
+        if (new_results == m_result_cache.end())
+            return;
+
+        // FIXME: Sort results based on weight of which would be best
+        on_new_results(new_results->value);
+    }
+
+    AppState& m_state;
+
     AppProvider m_app_provider;
     CalculatorProvider m_calculator_provider;
+    AsyncProvider m_async_provider;
 
-    // TODO: Cache results
-    Vector<NonnullRefPtr<Result>> m_results;
+    Threading::Lock m_lock;
+    HashMap<String, Vector<AK::NonnullRefPtr<Result>>> m_result_cache;
```

Comment:
`AK::` is not necessary unless in really rare cases for disambiguation.
all public AK classes/structs are imported into the global namespace.